src/storage/FileDescriptor.cpp

```diff
@@ -178,76 +180,180 @@ FileDescriptor FileDescriptor::stderrFactory()
     return FileDescriptor::stdFdFactory(STDERR_FILENO, "/dev/stderr");
 }
 
-FileDescriptor::FileDescriptor()
-  : iterStarted(false)
-  , iterFinished(false)
-  , dirPtr(nullptr)
-  , rightsSet(false)
-  , linuxFd(-1)
-  , linuxMode(-1)
-  , linuxFlags(-1)
-  , linuxErrno(0)
-  , wasiErrno(0)
-{}
-
 void FileDescriptor::iterReset()
 {
     // Reset iterator state
-    dirPtr = nullptr;
-    iterStarted = false;
-    iterFinished = false;
+    dirContentsLoaded = false;
+    dirContentsIdx = 0;
+    dirContents.clear();
 }
 
-DirEnt FileDescriptor::iterNext()
+void FileDescriptor::loadDirContents()
```

Comment:
The diff here is quite hard to read. I've basically split this up into a function which prefetches the whole directory (`loadDirContents`), and a much simpler one that handles the iteration (`iterNext`)